hoot-core/src/main/cpp/hoot/core/io/OsmApiChangeset.cpp

```diff
@@ -692,8 +692,10 @@ bool XmlChangeset::addRelation(ChangesetInfoPtr& changeset, ChangesetType type,
             //  Make sure that the ID is negative (create) in the ID map
             if (_idMap.getNewId(ElementType::Relation, member.getRef()) < 0)
             {
-              XmlRelation* relation = dynamic_cast<XmlRelation*>(_allRelations[member.getRef()].get());
-              addRelation(changeset, type, relation);
+              XmlRelation* relation_member = dynamic_cast<XmlRelation*>(_allRelations[member.getRef()].get());
+              //  Don't re-add self referencing relations
+              if (relation->id() != relation_member->id())
```

Comment:
All of the changes are just like this, if the relation's ID and the member's ID are the same then don't continue calling recursively.